vscode/src/extension.ts

```diff
@@ -145,14 +150,33 @@ function registerDocumentUpdateHandlers(languageService: ILanguageService) {
     }),
   );
 
-  function updateIfQsharpDocument(document: vscode.TextDocument) {
+  let currentlyCompiling = false;
+  let pendingCompile: vscode.TextDocument | null = null;
```

Comment:
Hmm, I'd definitely want to push this logic down to the Rust layer, leaving this code mostly as is. Basically, these bools can live in `language_service/src/lib.rs` instead. When the code is pending a compile, the language service requests (getCompletions etc) could yield empty results. The "recompile if there are pending changes" behavior can also be handled in the Rust layer.

This would keep the complexity in the language service itself, where it's more appropriate. It would also make the service behave correctly without requiring each client (vscode, playground, and maybe jupyterlab/lsp clients one day) to each implement their own logic.